compiler/fir/resolve/src/org/jetbrains/kotlin/fir/types/ConeTypeCompatibilityChecker.kt

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.types
+
+import org.jetbrains.kotlin.descriptors.Modality
+import org.jetbrains.kotlin.fir.declarations.*
+import org.jetbrains.kotlin.fir.isPrimitiveType
+import org.jetbrains.kotlin.fir.resolve.getSymbolByLookupTag
+import org.jetbrains.kotlin.fir.resolve.symbolProvider
+import org.jetbrains.kotlin.fir.symbols.ConeClassLikeLookupTag
+import org.jetbrains.kotlin.fir.symbols.ConeTypeParameterLookupTag
+import org.jetbrains.kotlin.fir.symbols.impl.FirTypeParameterSymbol
+import org.jetbrains.kotlin.name.StandardClassIds
+import org.jetbrains.kotlin.types.Variance
+import org.jetbrains.kotlin.utils.keysToMap
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Checks if a given collection of [ConeKotlinType] are compatible. In other words, the types are compatible if it's possible at all to
+ * define a type that's a subtype of all of the given types. The compatibility of a given set of types concept is closely related to whether
+ * the intersection of these types is inhabited. But it's not identical because 1) one can manually control visibility of constructors and
+ * 2) there can be unused type parameters.
+ *
+ * The compatibility check is done recursively on the given types and all type arguments passed to each corresponding type parameters. For
+ * example, consider the following two types:
+ *
+ * ```
+ *   - ArrayList<Set<String>>
+ *   - List<HashSet<Int>>
+ * ```
+ *
+ * The checker first checks the base types `ArrayList` and `List`, and it sees no issue since `ArrayList <: List`. Next it checks the type
+ * parameters bound to these base types: `T1` in `ArrayList<T1>` and `T2` in `List<T2>`. For `T1`, there is only one bound type argument
+ * `Set<String>`, so it's good. For `T2`, there are two bound type arguments: `Set<String>` and `HashSet<Int>`. Now the checker recursively
+ * checks whether these two types are compatible. Again, it first checks the base type `Set` and `HashSet`, and it finds no problem since
+ * `HashSet <: Set`. Finally, checks the type arguments `String` and `Int` that are bound to `T` in `Set<T>`. They are incompatible since
+ * `String` and `Int` are unrelated classes.
+ *
+ * The above example only goes over covariant type arguments. For contravariant types, the checker simply checks whether the range formed by
+ * covariant and contravariant bounds is empty. For example, a range like `[Collection, List]` is empty and hence invalid because `List` is
+ * not a super class/interface of `Collection`
+ */
+object ConeTypeCompatibilityChecker {
+
+    enum class Compatibility : Comparable<Compatibility> {
```

Comment:
I think you should mention that the order (ordinal) matters, since the notion of "upper bound" of compatibility is used below.